// @videogular/ngx-videogular/esm2022/core/lib/services/events/vg-events.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
class VgEvents {
    static { this.VG_ABORT = 'abort'; }
    static { this.VG_CAN_PLAY = 'canplay'; }
    static { this.VG_CAN_PLAY_THROUGH = 'canplaythrough'; }
    static { this.VG_DURATION_CHANGE = 'durationchange'; }
    static { this.VG_EMPTIED = 'emptied'; }
    static { this.VG_ENCRYPTED = 'encrypted'; }
    static { this.VG_ENDED = 'ended'; }
    static { this.VG_ERROR = 'error'; }
    static { this.VG_LOADED_DATA = 'loadeddata'; }
    static { this.VG_LOADED_METADATA = 'loadedmetadata'; }
    static { this.VG_LOAD_START = 'loadstart'; }
    static { this.VG_PAUSE = 'pause'; }
    static { this.VG_PLAY = 'play'; }
    static { this.VG_PLAYING = 'playing'; }
    static { this.VG_PROGRESS = 'progress'; }
    static { this.VG_RATE_CHANGE = 'ratechange'; }
    static { this.VG_SEEK = 'seek'; }
    static { this.VG_SEEKED = 'seeked'; }
    static { this.VG_SEEKING = 'seeking'; }
    static { this.VG_STALLED = 'stalled'; }
    static { this.VG_SUSPEND = 'suspend'; }
    static { this.VG_TIME_UPDATE = 'timeupdate'; }
    static { this.VG_VOLUME_CHANGE = 'volumechange'; }
    static { this.VG_WAITING = 'waiting'; }
    static { this.VG_LOAD = 'load'; }
    static { this.VG_ENTER = 'enter'; }
    static { this.VG_EXIT = 'exit'; }
    static { this.VG_START_ADS = 'startads'; }
    static { this.VG_END_ADS = 'endads'; }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VgEvents, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    /** @nocollapse */ static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VgEvents, providedIn: 'root' }); }
}
export { VgEvents };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VgEvents, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
