// @videogular/ngx-videogular/esm2022/core/lib/core.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgApiService } from './services/vg-api/vg-api.service';
import { VgControlsHiddenService } from './services/vg-controls-hidden/vg-controls-hidden.service';
import { VgFullscreenApiService } from './services/vg-fullscreen-api/vg-fullscreen-api.service';
import { VgUtilsService } from './services/vg-utils/vg-utils.service';
import { VgEvents } from './services/events/vg-events.service';
import { VgStates } from './services/states/vg-states.service';
import { VgCuePointsDirective } from './directives/vg-cue-points/vg-cue-points.directive';
import { VgMediaDirective } from './directives/vg-media/vg-media.directive';
import { VgPlayerComponent } from './components/vg-player/vg-player.component';
import * as i0 from "@angular/core";
const services = [
    VgApiService,
    VgControlsHiddenService,
    VgFullscreenApiService,
    VgUtilsService,
    VgEvents,
    VgStates
];
const directives = [
    VgCuePointsDirective,
    VgMediaDirective
];
class VgCoreModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VgCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.1", ngImport: i0, type: VgCoreModule, declarations: [VgCuePointsDirective,
            VgMediaDirective, VgPlayerComponent], imports: [CommonModule], exports: [VgCuePointsDirective,
            VgMediaDirective, VgPlayerComponent] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VgCoreModule, providers: [...services], imports: [CommonModule] }); }
}
export { VgCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: VgCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    providers: [...services],
                    declarations: [...directives, VgPlayerComponent],
                    exports: [...directives, VgPlayerComponent]
                }]
        }] });
